cmd/extensions/main.go

```diff
@@ -0,0 +1,449 @@
+	clientConf, err := clientcmd.BuildConfigFromFlags("", ctlConf.KubeConfig)
+	if err != nil {
+		logger.WithError(err).Fatal("Could not create in cluster config")
+	}
+
+	clientConf.QPS = float32(ctlConf.APIServerSustainedQPS)
+	clientConf.Burst = ctlConf.APIServerBurstQPS
+
+	kubeClient, err := kubernetes.NewForConfig(clientConf)
+	if err != nil {
+		logger.WithError(err).Fatal("Could not create the kubernetes clientset")
+	}
+
+	extClient, err := extclientset.NewForConfig(clientConf)
+	if err != nil {
+		logger.WithError(err).Fatal("Could not create the api extension clientset")
+	}
+
+	agonesClient, err := versioned.NewForConfig(clientConf)
+	if err != nil {
+		logger.WithError(err).Fatal("Could not create the agones api clientset")
+	}
+
+	cloudProduct, err := cloudproduct.New(ctx, ctlConf.CloudProduct, kubeClient)
+	if err != nil {
+		logger.WithError(err).Fatal("Could not initialize cloud provider")
+	}
+
+	// https server and the items that share the Mux for routing
+	httpsServer := https.NewServer(ctlConf.CertFile, ctlConf.KeyFile)
+	wh := webhooks.NewWebHook(httpsServer.Mux)
+	api := apiserver.NewAPIServer(httpsServer.Mux)
+
+	agonesInformerFactory := externalversions.NewSharedInformerFactory(agonesClient, defaultResync)
+	kubeInformerFactory := informers.NewSharedInformerFactory(kubeClient, defaultResync)
+
+	server := &httpServer{}
+	var rs []runner
+	var health healthcheck.Handler
+
+	// Stackdriver metrics
+	if ctlConf.Stackdriver {
+		sd, err := metrics.RegisterStackdriverExporter(ctlConf.GCPProjectID, ctlConf.StackdriverLabels)
+		if err != nil {
+			logger.WithError(err).Fatal("Could not register stackdriver exporter")
+		}
+		// It is imperative to invoke flush before your main function exits
+		defer sd.Flush()
+	}
+
+	// Prometheus metrics
+	if ctlConf.PrometheusMetrics {
+		registry := prom.NewRegistry()
+		metricHandler, err := metrics.RegisterPrometheusExporter(registry)
+		if err != nil {
+			logger.WithError(err).Fatal("Could not register prometheus exporter")
+		}
+		server.Handle("/metrics", metricHandler)
+		health = healthcheck.NewMetricsHandler(registry, "agones")
+	} else {
+		health = healthcheck.NewHandler()
+	}
+
+	// If we are using Prometheus only exporter we can make reporting more often,
+	// every 1 seconds, if we are using Stackdriver we would use 60 seconds reporting period,
+	// which is a requirements of Stackdriver, otherwise most of time series would be invalid for Stackdriver
+	metrics.SetReportingPeriod(ctlConf.PrometheusMetrics, ctlConf.Stackdriver)
+
+	// Add metrics controller only if we configure one of metrics exporters
+	if ctlConf.PrometheusMetrics || ctlConf.Stackdriver {
+		rs = append(rs, metrics.NewController(kubeClient, agonesClient, kubeInformerFactory, agonesInformerFactory))
+	}
+
+	server.Handle("/", health)
+
+	gsCounter := gameservers.NewPerNodeCounter(kubeInformerFactory, agonesInformerFactory)
+
+	gsController := gameservers.NewController(wh, health,
```

Comment:
Unless I'm missing something, this does mean we are running a lot of controllers at the same time for the same resources, just so I'm clear - is that something we are concerned with at this stage, or is this just a pre-phase of https://github.com/googleforgames/agones/issues/2797#user-content-phase1?

Reply:
Exactly the latter. The point of this PR was that I could check it out and review that there were only superficial deltas. After, the changes to main.go will be more obvious - GitHub doesn't show cloned files very well.

Reply:
Sorry, I think I'm having a 5pm dead brain moment. Not sure I understand.

> Exactly the latter. 

When you say latter, do you mean I am correct that this is:
> ...this just a pre-phase of [#2797 (comment)](https://github.com/googleforgames/agones/issues/2797#user-content-phase1)?

> ...review that there were only superficial deltas.

Superficial deltas to what?

Reply:
Yeah, this is a pre-phase, because..

> Superficial deltas to what?

GitHub doesn't make reviewing cloned files easy, so for this one I checked the PR out and reviewed like:

```
$ diff -U3 cmd/controller cmd/extensions
Only in cmd/controller: bin
diff '--color=auto' -U3 cmd/controller/Dockerfile cmd/extensions/Dockerfile
--- cmd/controller/Dockerfile	2022-10-04 16:23:39.811044734 +0000
+++ cmd/extensions/Dockerfile	2022-12-15 01:05:53.706043689 +0000
@@ -18,8 +18,8 @@
 RUN apk --update add ca-certificates && \
     adduser -D -u 1000 agones

-COPY --chown=agones:agones ./bin/controller.linux.$ARCH /home/agones/controller
+COPY --chown=agones:agones ./bin/extensions.linux.$ARCH /home/agones/extensions
 COPY --chown=agones:agones ./bin/LICENSES ./bin/dependencies-src.tgz /home/agones/

 USER 1000
-ENTRYPOINT ["/home/agones/controller"]
+ENTRYPOINT ["/home/agones/extensions"]
diff '--color=auto' -U3 cmd/controller/main.go cmd/extensions/main.go
--- cmd/controller/main.go	2022-12-14 23:58:58.116573574 +0000
+++ cmd/extensions/main.go	2022-12-15 01:05:53.706043689 +0000
@@ -88,7 +88,7 @@
 )

 func setupLogging(logDir string, logSizeLimitMB int) {
-	logFileName := filepath.Join(logDir, "agones-controller-"+time.Now().Format("20060102_150405")+".log")
+	logFileName := filepath.Join(logDir, "agones-extensions-"+time.Now().Format("20060102_150405")+".log")

 	const maxLogSizeMB = 100
 	maxBackups := (logSizeLimitMB - maxLogSizeMB) / maxLogSizeMB
@@ -130,7 +130,7 @@
 		for _, err := range errs {
 			logger.Error(err)
 		}
-		logger.Fatal("Could not create controller from environment or flags")
+		logger.Fatal("Could not create extensions from environment or flags")
 	}

 	// if the kubeconfig fails BuildConfigFromFlags will try in cluster config
@@ -239,7 +239,7 @@
 	}

 	<-ctx.Done()
-	logger.Info("Shut down agones controllers")
+	logger.Info("Shut down agones extensions")
 }

 func parseEnvFlags() config {
```

Those were the "superficial diffs". If there were many more diffs, it'd be more painful. The next PRs can be more easily reviewed/commented in GitHub.

Reply:
Gotcha. That makes sense - just wanted to be sure / make sure there was a decision audit trail as well. 👍🏻

Outside of nits, this looks good to me.